tempesta_fw/str_avx2.S

```diff
@@ -0,0 +1,2296 @@
+.str2low_len6:
+	__STRTOLOWER_SMALL_STR 5
+.str2low_len5:
+	__STRTOLOWER_SMALL_STR 4
+.str2low_len4:
+	__STRTOLOWER_SMALL_STR 3
+.str2low_len3:
+	__STRTOLOWER_SMALL_STR 2
+.str2low_len2:
+	__STRTOLOWER_SMALL_STR 1
+.str2low_len1:
+	__STRTOLOWER_SMALL_STR 0
+.str2low_len0:
+	addq	$88, %rsp
+	popq	%rbx
+	popq	%r10
+	popq	%r12
+	popq	%rbp
+	leaq	-8(%r10), %rsp
+	ret
+
+	/*
+	 * Place the processing code before the conditions block to avoid 2
+	 * jumps: for the 128 byte processing loop and following 64 byte block.
+	 */
+.str2low_tolower128:
+	movslq	%r10d, %rax
+	vmovdqa	__A, %ymm4
+	leal	-128(%rax), %r10d
+	vmovdqa	__D, %ymm2
+	vmovdqa	__CASE, %ymm0
+	movslq	%r10d, %r10
+	leaq	(%rsi,%r10), %r11
+	addq	%rdi, %r10
+	vlddqu	(%r11), %ymm7
+	vlddqu	32(%r11), %ymm6
+	vpsubb	%ymm4, %ymm7, %ymm1
+	vpcmpgtb %ymm1, %ymm2, %ymm1
+	vlddqu	64(%r11), %ymm5
+	vlddqu	96(%r11), %ymm3
+	vpand	%ymm1, %ymm0, %ymm1
+	vpor	%ymm7, %ymm1, %ymm1
+	vmovdqu	%ymm1, (%r10)
+	vpsubb	%ymm4, %ymm6, %ymm1
+	vpcmpgtb %ymm1, %ymm2, %ymm1
+	vpand	%ymm1, %ymm0, %ymm1
+	vpor	%ymm6, %ymm1, %ymm1
+	vmovdqu	%ymm1, 32(%r10)
+	vpsubb	%ymm4, %ymm5, %ymm1
+	vpsubb	%ymm4, %ymm3, %ymm4
+	vpcmpgtb %ymm1, %ymm2, %ymm1
+	vpcmpgtb %ymm4, %ymm2, %ymm2
+	vpand	%ymm1, %ymm0, %ymm1
+	vpand	%ymm2, %ymm0, %ymm0
+	vpor	%ymm5, %ymm1, %ymm1
+	vpor	%ymm3, %ymm0, %ymm0
+	vmovdqu	%ymm1, 64(%r10)
+	vmovdqu	%ymm0, 96(%r10)
+
+.str2low_test_len128:
+	__STRTOLOWER_TEST_LEN 128 %r10d %r11 .str2low_tolower128
+	__STRTOLOWER_TEST_LEN 64 %r10d %r11 .str2low_tolower64
+	movslq	%eax, %r11
+.str2low_test_len32:
+	__STRTOLOWER_TEST_LEN 32 %r10d %rbx .str2low_tolower32
+.str2low_test_len16:
+	__STRTOLOWER_TEST_LEN 16 %r10d %rbx .str2low_tolower16
+.str2low_test_len8:
+	__STRTOLOWER_TEST_LEN 8 %ebx %r10 .str2low_tolower8
+	movq	%r11, %r10
+.str2low_small_len:
+	subq	%r10, %rdx
+	addq	%r10, %rdi
+	addq	%r10, %rsi
+	jmp	.str2low_prefix_suffix
+
+.str2low_tolower64:
+	leaq	(%rsi,%rax), %rbx
+	vmovdqa	__A, %ymm3
+	addq	%rdi, %rax
+	vmovdqa	__D, %ymm2
+	vlddqu	(%rbx), %ymm5
+	vmovdqa	__CASE, %ymm1
+	vpsubb	%ymm3, %ymm5, %ymm0
+	vpcmpgtb %ymm0, %ymm2, %ymm0
+	vlddqu	32(%rbx), %ymm4
+	vpand	%ymm0, %ymm1, %ymm0
+	vpor	%ymm5, %ymm0, %ymm0
+	vmovdqu	%ymm0, (%rax)
+	vpsubb	%ymm3, %ymm4, %ymm0
+	vpcmpgtb %ymm0, %ymm2, %ymm0
+	vpand	%ymm0, %ymm1, %ymm0
+	vpor	%ymm4, %ymm0, %ymm0
+	vmovdqu	%ymm0, 32(%rax)
+	movslq	%r10d, %rax
+	jmp	.str2low_test_len32
+
+.str2low_tolower32:
+	vlddqu	(%rsi,%r11), %ymm1
+	movslq	%r10d, %rax
+	vmovdqa	__D, %ymm0
+	vpsubb	__A, %ymm1, %ymm2
+	vpcmpgtb %ymm2, %ymm0, %ymm0
+	vpand	__CASE, %ymm0, %ymm0
+	vpor	%ymm1, %ymm0, %ymm0
+	vmovdqu	%ymm0, (%rdi,%r11)
+	movq	%rbx, %r11
+	jmp	.str2low_test_len16
+
+.str2low_tolower16:
+	vlddqu	(%rsi,%r11), %xmm1
+	movslq	%r10d, %rax
+	vmovdqa	__D, %xmm0
+	vpsubb	__C(%rip), %xmm1, %xmm2
+	vpcmpgtb %xmm2, %xmm0, %xmm0
+	vpand	__CASE, %xmm0, %xmm0
+	vpor	%xmm1, %xmm0, %xmm0
+	vmovups	%xmm0, (%rdi,%r11)
+	movq	%rbx, %r11
+	jmp	.str2low_test_len8
+
+.str2low_tolower8:
```

Comment:
Use of `volatile` in the C source code resulted in obviously redundant memory stores and loads. As far as I understand, code below is equivalent to:
```asm
        movq    (%rsi,%r11), %rax
        movq    %r9, %mm2
        movq    %r8, %mm1
        movq    %rax, %mm0
        movq    %rax, %r12
        psubb   %mm2, %mm0
        pcmpgtb %mm0, %mm1
        movq    %mm1, %rax
        andq    %rcx, %rax
        orq     %r12, %rax
        movq    %rax, (%rdi,%r11)
        movslq  %ebx, %rax
        jmp     .str2low_small_len
```